dali/kernels/audio/mel_scale/mel_filter_bank_gpu.cu

```diff
@@ -93,28 +105,225 @@ __global__ void MelFilterBankKernel(const BlockDesc<T> *block_desc,
-  T *out =  block_desc[block_id].out_frame;
-  T norm_factor = (normalize) ? norm_factors[mel_bin] : 1;
-  *(out + idx) = calcMel(in + window * fftdim, mel_bin,
-                         weights_down, interval_ends, 1, 0, norm_factor);
+template <typename T>
+DALI_HOST_DEV
+constexpr int shm_out_size(int shm_height, int mel) {
+  return shm_out_stride<T>(mel) * shm_height * sizeof(T);
 }
 
+/**
+ * @brief Implements MelFilterbank with inner FFT
+ *
+ * @tparam shm_height the number of data rows stored in shared memory; must be 1, 2, 4, 8, 16 or 32
+ * @tparam T the data type (float or double)
+ */
+template <int shm_height, typename T>
+struct MelFilterBankInnerFft {
+  BlockDesc<T> block_desc;
+  const T *__restrict__ weights_down;
+  const T *__restrict__ weights_up;
+  const int *__restrict__ bin_down;
+  int fft_lo, fft_hi;
+  int nmel, nfft;
+  char *shm;
+
+  static const int max_freqs = kMaxInnerFftFreqs;
+  static const int sh_in_stride = max_freqs + 1;
+  int sh_out_stride = 0;
+
+  __device__ T &shm_in(int window, int freq) const {
+    return reinterpret_cast<T*>(shm)[sh_in_stride * window + freq];
+  }
+
+  __device__ T &shm_out(int window, int mel) const {
+    const int shm_out_offset = shm_in_size<T>(shm_height);
+    return reinterpret_cast<T*>(shm + shm_out_offset)[window * sh_out_stride + mel];
+  }
+
+  __device__ void Run() {
+    sh_out_stride = shm_out_stride<T>(nmel);
+
+    bool first = true;
+
+    for (int start_window = 0;
+         start_window < block_desc.frame_nwindows;
+         start_window += shm_height) {
+      if (first)
+        first = false;
+      else
+        __syncthreads();
+
+      int fft_end;
+      // The inner loop accumulates Mel coefficients from FFT spectrum
+      ClearMel();
+      for (int fft_start = fft_lo; fft_start < fft_hi; fft_start = fft_end) {
+        // avoid last thin slice - use up to max_freqs witdth, if it fits
+        if (fft_hi - fft_start <= max_freqs)
+          fft_end = fft_hi;
+        else
+          fft_end = fft_start + 32;
+        // Load a block of FFT coefficients into shared memory
+        LoadFrequencyBlock(start_window, fft_start, fft_end);
+        __syncthreads();
+        // Go over FFT coefficients and accumulate Mel values in shared memory
+        ProcessFrequencyBlock(fft_start, fft_end);
+        __syncthreads();
+      }
+      // Store the final Mel values in global memory
+      StoreMel(start_window);
+    }
+  }
+
+  __device__ void LoadFrequencyBlock(int start_window, int fft_start, int fft_end) {
+    int bh = blockDim.y;
+    const T *in = block_desc.in_frame;
+    // block height might be smaller than shm_height, so we need to iterate
+    for (int y = threadIdx.y; y < shm_height; y += bh) {
+      int window = start_window + y;
+      if (window < block_desc.frame_nwindows) {
+        for (int x = threadIdx.x; x < fft_end - fft_start; x += 32)
+          shm_in(y, x) = in[window * nfft + x + fft_start];
```

Comment:
`in + window * nfft + fft_start` could be calculated outside of the inner loop once.

Reply:
Sure, but given how short the loop body is, I doubt the compiler would be stupid enough not to hoist it.